### PR TITLE
feat: add fluxv2 addons

### PR DESCRIPTION
**Pull request by rguichard**

### CODE REVIEW

**Comment by ArchiFleKs** [comment]
Could you implement the `create_ns` logic on new addition like [here](https://github.com/particuleio/terraform-kubernetes-addons/blob/main/karma.tf#L9) and also the label monitoring to not depend on `monitoring` namespace creation like [here]()

File: fluxv2.tf [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    ingress {
+      ports {
+        port     = "3030"
+        protocol = "TCP"
+      }
+
+      from {
+        namespace_selector {
+          match_labels = {
+            name = kubernetes_namespace.kube-prometheus-stack.*.metadata.0.name[count.index]
```
Comment (ArchiFleKs) [inline fluxv2.tf:187]: `"${local.labels_prefix}/component" = "monitoring"`

File: fluxv2.tf [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  content    = data.flux_sync.main[0].kustomize_content
+  branch     = local.fluxv2["branch"]
+}
+
+resource "kubernetes_network_policy" "fluxv2_allow_monitoring" {
+  count = local.fluxv2["enabled"] && local.fluxv2["network_policy"] && local.kube-prometheus-stack["enabled"] ? 1 : 0
+
+  metadata {
+    name      = "${kubernetes_namespace.fluxv2.*.metadata.0.name[count.index]}-allow-monitoring"
+    namespace = kubernetes_namespace.fluxv2.*.metadata.0.name[count.index]
```
Comment (ArchiFleKs) [inline fluxv2.tf:171]:
```
    name      = "${local.fluxv2["create_ns"] ? kubernetes_namespace.fluxv2.*.metadata.0.name[count.index] : local.fluxv2["namespace"]}-allow-ingress-fluxv2"
    namespace = local.fluxv2["create_ns"] ? kubernetes_namespace.fluxv2.*.metadata.0.name[count.index] : local.fluxv2["namespace"]
```

File: fluxv2.tf [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+resource "github_repository_file" "kustomize" {
+  count      = local.fluxv2["enabled"] ? 1 : 0
+  repository = github_repository.main[0].name
+  file       = data.flux_sync.main[0].kustomize_path
+  content    = data.flux_sync.main[0].kustomize_content
+  branch     = local.fluxv2["branch"]
+}
+
+resource "kubernetes_network_policy" "fluxv2_allow_monitoring" {
+  count = local.fluxv2["enabled"] && local.fluxv2["network_policy"] && local.kube-prometheus-stack["enabled"] ? 1 : 0
```
Comment (ArchiFleKs) [inline fluxv2.tf:167]:
```
count = local.fluxv2["enabled"] && local.fluxv2["network_policy"]
```

File: fluxv2.tf [hunk 4]
```diff
@@ -0,0 +1,195 @@
+locals {
+
+  known_hosts = "github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
+  fluxv2 = merge(
+    {
+      enabled               = false
+      create_ns             = true
+      namespace             = "flux-system"
+      target_path           = "production"
+      network_policy        = true
```
Comment (ArchiFleKs) [inline fluxv2.tf:10]: `default_network_policy`

**Review by ArchiFleKs — CHANGES_REQUESTED** [review]
It is also missing the default policy that allow its own namespace like in the other files

File: fluxv2.tf [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      version               = "v0.8.0"
+      github_url            = "https://github.com/marie/curie"
+      repository            = "curie"
+      repository_visibility = "public"
+      branch                = "main"
+      github_token          = var.fluxv2["github_token"]
```
Comment (ArchiFleKs) [inline fluxv2.tf:16]: set to `null` or `""` depending on what works with validation, it will be set with the merge below if set in `var.fluxv2`

**Review by ArchiFleKs — CHANGES_REQUESTED** [review]

**Comment by ArchiFleKs** [comment]
Finally, I think you should `s/fluxv2/flux2` :p

File: fluxv2.tf (hunk 3 above)
Comment (rguichard) [reply fluxv2.tf:167]: `network_policy` refers to https://registry.terraform.io/providers/fluxcd/flux/latest/docs/data-sources/install
Should we  `s/network_policy/default_network_policy/`everywhere even if it's not the correct name in the datasource ?

File: fluxv2.tf (hunk 2 above)
Comment (rguichard) [reply fluxv2.tf:171]: Whether `create_ns` is used or not, the name and the namespace will always be the same so we could just use `local.flux2["namespace]` as metadata name and metadata namespace.

File: fluxv2.tf (hunk 4 above)
Comment (rguichard) [reply fluxv2.tf:10]: Same as above

**Comment by rguichard** [comment]
regarding 
```
resource "kubectl_manifest" "sync" {                                             
  for_each   = { for v in local.sync : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
  depends_on = [kubernetes_namespace.flux2]                                      
  yaml_body  = each.value                                                        
}    
```
If we use that, this resource will not be created when `create_ns = false` because the namespace resource won't exist isn't it ?

File: fluxv2.tf (hunk 2 above)
Comment (ArchiFleKs) [reply fluxv2.tf]: yes but if you create the namespace this create an implicit dependency so terraform does not try to deploy the chart before the NS, it is needed when you create the namespace

File: fluxv2.tf (hunk 3 above)
Comment (ArchiFleKs) [reply fluxv2.tf]: ok but the network policy is handled with flux manifest right ? not with terraform ? the `default_network_policy` refers to the `default_network_policy` deploy with TF in all other addons

File: fluxv2.tf (hunk 4 above)
Comment (ArchiFleKs) [reply fluxv2.tf]: If this is a parameter needed by flux, we should have both, if flux deploy its own network policy. But keep our terraform defined policies

**Comment by ArchiFleKs** [comment]
> regarding
> 
> ```
> resource "kubectl_manifest" "sync" {                                             
>   for_each   = { for v in local.sync : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
>   depends_on = [kubernetes_namespace.flux2]                                      
>   yaml_body  = each.value                                                        
> }    
> ```
> 
> If we use that, this resource will not be created when `create_ns = false` because the namespace resource won't exist isn't it ?

you mean regarding the `depends_on`?

File: fluxv2.tf (hunk 3 above)
Comment (rguichard) [reply fluxv2.tf]: You should disabled `network_policy` and handle our own netpol with a terraform resource

**Comment by rguichard** [comment]
> > regarding
> > ```
> > resource "kubectl_manifest" "sync" {                                             
> >   for_each   = { for v in local.sync : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
> >   depends_on = [kubernetes_namespace.flux2]                                      
> >   yaml_body  = each.value                                                        
> > }    
> > ```
> > 
> > 
> > If we use that, this resource will not be created when `create_ns = false` because the namespace resource won't exist isn't it ?
> 
> you mean regarding the `depends_on`?

yes

**Comment by ArchiFleKs** [comment]
> > > regarding
> > > ```
> > > resource "kubectl_manifest" "sync" {                                             
> > >   for_each   = { for v in local.sync : lower(join("/", compact([v.data.apiVersion, v.data.kind, lookup(v.data.metadata, "namespace", ""), v.data.metadata.name]))) => v.content }
> > >   depends_on = [kubernetes_namespace.flux2]                                      
> > >   yaml_body  = each.value                                                        
> > > }    
> > > ```
> > > 
> > > 
> > > If we use that, this resource will not be created when `create_ns = false` because the namespace resource won't exist isn't it ?
> > 
> > 
> > you mean regarding the `depends_on`?
> 
> yes

Depends_on are not used if the resource is not created like [here](https://github.com/particuleio/terraform-kubernetes-addons/blob/main/modules/aws/cert-manager.tf#L146) it works even when kube-prometheus-stack is not enabled

File: flux2.tf [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    policy_types = ["Ingress"]
+  }
+}
+
+resource "kubernetes_network_policy" "flux2_allow_namespace" {
+  count = local.flux2["enabled"] && local.flux2["default_network_policy"] ? 1 : 0
+
+  metadata {
+    name      = "${local.flux2["create_ns"] ? kubernetes_namespace.flux2.*.metadata.0.name[count.index] : local.flux2["namespace"]}-allow-ingress-flux2"
```
Comment (ArchiFleKs) [inline flux2.tf:202]: just `-allow-namespace`

**Review by ArchiFleKs — CHANGES_REQUESTED** [review]

**Review by ArchiFleKs — APPROVED** [review]

**Pull request merged**